server/src/main/java/org/opensearch/index/ShardIndexingPressure.java

```diff
@@ -0,0 +1,334 @@
+        long nodeTotalBytes = nodeCombinedBytes + nodeReplicaBytes;
+        long shardCombinedBytes = tracker.getCommonOperationTracker().incrementCurrentCombinedCoordinatingAndPrimaryBytes(bytes);
+
+        boolean shardLevelLimitBreached = false;
+        if (forceExecution == false) {
+            boolean nodeLevelLimitBreached = memoryManager.isCoordinatingNodeLimitBreached(tracker, nodeTotalBytes);
+            if (nodeLevelLimitBreached == false) {
+                shardLevelLimitBreached = memoryManager.isCoordinatingShardLimitBreached(tracker, nodeTotalBytes, requestStartTime);
+            }
+
+            if (shouldRejectRequest(nodeLevelLimitBreached, shardLevelLimitBreached)) {
+                coordinatingRejections.getAndIncrement();
+                currentCombinedCoordinatingAndPrimaryBytes.addAndGet(-bytes);
+                tracker.getCommonOperationTracker().incrementCurrentCombinedCoordinatingAndPrimaryBytes(-bytes);
+                rejectShardRequest(tracker, bytes, nodeTotalBytes, shardCombinedBytes,
+                    tracker.getCoordinatingOperationTracker().getRejectionTracker(), "coordinating");
+            }
+        }
+        currentCoordinatingBytes.addAndGet(bytes);
+        totalCombinedCoordinatingAndPrimaryBytes.addAndGet(bytes);
+        totalCoordinatingBytes.addAndGet(bytes);
+
+        StatsTracker statsTracker = tracker.getCoordinatingOperationTracker().getStatsTracker();
+        statsTracker.incrementCurrentBytes(bytes);
+        markShardOperationStarted(statsTracker, tracker.getCoordinatingOperationTracker().getPerformanceTracker());
+        boolean isShadowModeBreach = shardLevelLimitBreached;
+
+        return () -> {
+            currentCombinedCoordinatingAndPrimaryBytes.addAndGet(-bytes);
+            currentCoordinatingBytes.addAndGet(-bytes);
+            markShardOperationComplete(bytes, requestStartTime, isShadowModeBreach, tracker.getCoordinatingOperationTracker(),
+                tracker.getCommonOperationTracker());
+            memoryManager.decreaseShardPrimaryAndCoordinatingLimits(tracker);
+            tryReleaseTracker(tracker);
+        };
+    }
+
+    public Releasable markPrimaryOperationLocalToCoordinatingNodeStarted(ShardId shardId, long bytes) {
+        if(bytes == 0) { return () -> {}; }
+
+        ShardIndexingPressureTracker tracker = getShardIndexingPressureTracker(shardId);
+
+        currentPrimaryBytes.addAndGet(bytes);
+        totalPrimaryBytes.addAndGet(bytes);
+        tracker.getPrimaryOperationTracker().getStatsTracker().incrementCurrentBytes(bytes);
+        tracker.getPrimaryOperationTracker().getStatsTracker().incrementTotalBytes(bytes);
+
+        return () -> {
+            currentPrimaryBytes.addAndGet(-bytes);
+            tracker.getPrimaryOperationTracker().getStatsTracker().incrementCurrentBytes(-bytes);
+        };
+    }
+
+    public Releasable markPrimaryOperationStarted(ShardId shardId, long bytes, boolean forceExecution) {
```

Comment:
It is known that the Releasable can be called multiple times. Should we wrap it to ensure it gets executed only once
```
    private static Releasable wrapReleasable(Releasable releasable) {
        final AtomicBoolean called = new AtomicBoolean();
        return () -> {
            if (called.compareAndSet(false, true)) {
                releasable.close();
            } else {
                logger.error("IndexingPressure memory is adjusted twice", new IllegalStateException("Releasable is called twice"));
                assert false : "IndexingPressure is adjusted twice";
            }
        };
    }
```

Reply:
In the original PR from elastic had no specific reason pointed yet, but more of a safety net. However, I agree having it wrapped with associated error log will help clients of ``ShardIndexingPressure`` to identify any leaks, and call Releasable at most once.